### ISSUE TITLE
staticEval is not static

### ISSUE THREAD

**Issue by vondele** [official-stockfish/Stockfish#1335]
Since the introduction of lazyEval, staticEval is not static anymore, but can depend on the path in the search graph. In particular, if a config is reached via a null move, it will have a different staticEval as when it is reached from a normal move (assuming they are in the lazyEval regime). The obvious fix 
https://github.com/vondele/Stockfish/commit/8503b7abb9e8fe1d68e3e2aeff79308421cbd146
was tested here: 
http://tests.stockfishchess.org/tests/view/5a2d57700ebc590ccbb8bb37
but didn't pass [-3,1]

**Comment by Stefano80** [comment]
Why is this an issue?

**Comment by vondele** [comment]
Because staticEval is the static evaluation of that position, irrespective of the path (contrary to eval). It seems quite a logical thing to even code an assert(staticEval == evaluate(pos)) to maintain such an internal consistency.

I would say that the current state is just an oversight of the fact that we compute
 ```
ss->staticEval =
        (ss-1)->currentMove != MOVE_NULL ? evaluate(pos)
                                         : -(ss-1)->staticEval + 2 * Eval::Tempo;
``` 
and not deliberate design.

**Comment by Stefano80** [comment]
It was definitely not an oversight, given that I happen to know this place in search.cpp by heart because of past experiments.

I think consistency really does not matter here, I am evaluating lazily...

**Comment by vondele** [comment]
I'd turn it around. I don't think there is a benefit in being inconsistent at this point, why not keep the code in a state that keeps it easier to reason about ?

**Comment by joergoster** [comment]
Please note, we also don't add Eval::Tempo in case a specialized eval exists, and yet the same staticEval trick is applied.

**Comment by vondele** [comment]
interesting, would also be inconsistent with the null move scoring trick. So, we maybe just rename staticEval to posEval ? At least that doesn't suggest it is static.

**Comment by vondele** [comment]
I've scheduled another try, which would catch both cases.

**Comment by joergoster** [comment]
Well, this is the commit where tempo bonus were removed from specialized eval.
https://github.com/official-stockfish/Stockfish/commit/9cae6e66ce00850086d1bfe1e24e34442c12b206

To be honest, I think it is the null move scoring trick that is causing the trouble here. Unfortunately, it seems to gain a small amount of Elo.
OTOH, is it really worth fixing this very little hickup?

**Comment by vondele** [comment]
OK, with the patch running here http://tests.stockfishchess.org/tests/view/5a2e893f0ebc590ccbb8bba3 staticEval == evaluation(pos) almost everywhere. Exceptions are hash table collisions. 

Just for fun, the following fens collide from depth 13 on the startpos with default (16Mb) hash:
```
Fen: rnbqkb1r/1pp1nppp/p7/3p4/3P4/2P5/PP2QPPP/RNB1KBNR w KQkq - 0 6
Key: C8CFF653E32A18BE (tt index 218BE, key16 C8CF)
evaluation=139

Fen: rnbqkb1r/pp3ppp/2p1p3/3P4/3Pn3/2N2N2/PPP2PPP/R1BQKB1R w KQkq - 0 6
Key: C8CF812F19C218BE (tt index 218BE, key16 C8CF)
evaluation=417
```

**Comment by mcostalba** [comment]
Closed after f6981b13213578 has been merged
